File: phone-app/src/services/pairing.ts
/**
 * pairing.ts — decode the one code that brings the trio together.
 *
 * Mirrors host-python/src/dreamlayer/pairing.py: a `dreamlayer:` deep-link
 * whose payload is URL-safe base64 of a tiny JSON bundle carrying the Brain
 * URL, the pairing token, and the glasses' BLE id. The Mac mini panel shows
 * it as a QR; the phone scans (or pastes) it and is instantly wired.
 *
 * We hand-roll base64 so this works on any RN/Hermes runtime without relying
 * on atob/btoa being present.
 */
export const SCHEME = "dreamlayer";

export type PairingBundle = {
  brainUrl: string;
  token: string;
  glassesId: string;
  label: string;
  relayUrl: string; // reach the Brain off your LAN (optional)
};

const B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

function bytesToUtf8(bytes: number[]): string {
  // minimal UTF-8 decode (the payload is JSON; usually ASCII)
  let out = "";
  for (let i = 0; i < bytes.length; ) {
    const b = bytes[i++];
    if (b < 0x80) out += String.fromCharCode(b);
    else if (b < 0xe0) out += String.fromCharCode(((b & 0x1f) << 6) | (bytes[i++] & 0x3f));
    else if (b < 0xf0)
      out += String.fromCharCode(((b & 0x0f) << 12) | ((bytes[i++] & 0x3f) << 6) | (bytes[i++] & 0x3f));
    else {
      const cp =
        ((b & 0x07) << 18) | ((bytes[i++] & 0x3f) << 12) | ((bytes[i++] & 0x3f) << 6) | (bytes[i++] & 0x3f);
      const c = cp - 0x10000;
      out += String.fromCharCode(0xd800 + (c >> 10), 0xdc00 + (c & 0x3ff));
    }
  }
  return out;
}

function utf8ToBytes(s: string): number[] {
  const out: number[] = [];
  for (const ch of s) {
    let c = ch.codePointAt(0)!;
    if (c < 0x80) out.push(c);
    else if (c < 0x800) out.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
    else if (c < 0x10000) out.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
    else out.push(0xf0 | (c >> 18), 0x80 | ((c >> 12) & 0x3f), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
  }
  return out;
}

function b64urlEncode(bytes: number[]): string {
  const std = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const n = (bytes[i] << 16) | ((bytes[i + 1] ?? 0) << 8) | (bytes[i + 2] ?? 0);
    out += std[(n >> 18) & 63] + std[(n >> 12) & 63];
    out += i + 1 < bytes.length ? std[(n >> 6) & 63] : "=";
    out += i + 2 < bytes.length ? std[n & 63] : "=";
  }
  return out.replace(/\+/g, "-").replace(/\//g, "_");
}

function b64urlDecode(s: string): number[] {
  const std = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  const clean = s.replace(/-/g, "+").replace(/_/g, "/").replace(/=+$/, "");
  const out: number[] = [];
  for (let i = 0; i < clean.length; i += 4) {
    const n =
      (std.indexOf(clean[i]) << 18) |
      (std.indexOf(clean[i + 1]) << 12) |
      ((clean[i + 2] ? std.indexOf(clean[i + 2]) : 0) << 6) |
      (clean[i + 3] ? std.indexOf(clean[i + 3]) : 0);
    out.push((n >> 16) & 0xff);
    if (clean[i + 2]) out.push((n >> 8) & 0xff);
    if (clean[i + 3]) out.push(n & 0xff);
  }
  return out;
}

export function encodePairing(b: PairingBundle): string {
  const payload: Record<string, string> = {};
  if (b.brainUrl) payload.brain_url = b.brainUrl;
  if (b.token) payload.token = b.token;
  if (b.glassesId) payload.glasses_id = b.glassesId;
  if (b.label && b.label !== "DreamLayer") payload.label = b.label;
  if (b.relayUrl) payload.relay_url = b.relayUrl;
  return SCHEME + ":" + b64urlEncode(utf8ToBytes(JSON.stringify(payload)));
}

export function decodePairing(code: string): PairingBundle {
  let s = code.trim();
  if (s.startsWith(SCHEME + ":")) s = s.slice(SCHEME.length + 1);
  const data = JSON.parse(bytesToUtf8(b64urlDecode(s))) as Record<string, string>;
  return {
    brainUrl: data.brain_url ?? "",
    token: data.token ?? "",
    glassesId: data.glasses_id ?? "",
    label: data.label ?? "DreamLayer",
    relayUrl: data.relay_url ?? "",
  };
}

// referenced so the alphabet constant isn't flagged unused by strict builds
void B64;

File: host-python/src/dreamlayer/pairing.py
"""pairing.py — connect phone + brain + glasses with one code.

Pairing should be a scan, not a setup wizard. The Mac mini Brain shows a
pairing code (the panel renders it as a QR); the phone reads it and is
instantly wired to the Brain. The same bundle can carry the glasses' BLE id
so one code brings the whole trio together.

    bundle = PairingBundle(brain_url="http://studio-mbp.local:7777",
                           token="rune-birch", glasses_id="HALO-9F2A")
    code = encode_pairing(bundle)          # → shown as a QR on the Brain panel
    # …phone scans…
    connect_all(orc, decode_pairing(code)) # phone now talks to Brain + glasses

The code is a compact, URL-safe base64 of a tiny JSON — no secrets beyond the
pairing token you already chose, and it only ever travels the way you show it
(a QR on your own screen).
"""
from __future__ import annotations

import base64
import binascii
import json
from dataclasses import dataclass, asdict

SCHEME = "dreamlayer"          # dreamlayer:<base64> for deep-link / QR

# A pairing code is a tiny JSON (a URL, a short token, a BLE id). Anything much
# larger is not ours — cap the input so a hostile QR/deep-link can't hand us a
# multi-megabyte blob to base64-decode + json-parse (audit 2026-07-14: decode
# had no size cap and no error handling).
_MAX_CODE_LEN = 4096
_OK_SCHEMES = ("http://", "https://")


@dataclass
class PairingBundle:
    brain_url: str = ""
    token: str = ""
    glasses_id: str = ""       # BLE identifier of the glasses (optional)
    label: str = "DreamLayer"
    relay_url: str = ""        # reach the Brain off your LAN (optional)


def encode_pairing(bundle: PairingBundle) -> str:
    payload = {k: v for k, v in asdict(bundle).items() if v}
    raw = json.dumps(payload, separators=(",", ":")).encode("utf-8")
    return SCHEME + ":" + base64.urlsafe_b64encode(raw).decode("ascii")


def _checked_url(value: str, field_name: str) -> str:
    """A decoded URL is fed into live HTTP — refuse anything that isn't http(s)
    so a crafted code can't smuggle a file:// / javascript: scheme into the
    connect path (audit 2026-07-14). Empty is allowed (the field is optional)."""
    if value and not value.startswith(_OK_SCHEMES):
        raise ValueError(f"pairing {field_name} must be http(s): {value!r}")
    return value


def decode_pairing(code: str) -> PairingBundle:
    """Decode a pairing code back to a bundle. Raises ``ValueError`` on anything
    malformed (bad base64, non-JSON, oversized, or a non-http(s) URL) so callers
    get one predictable exception type instead of a raw binascii/JSON error."""
    s = (code or "").strip()
    if len(s) > _MAX_CODE_LEN:
        raise ValueError(f"pairing code too long ({len(s)} > {_MAX_CODE_LEN})")
    if s.startswith(SCHEME + ":"):
        s = s[len(SCHEME) + 1:]
    try:
        raw = base64.urlsafe_b64decode(s.encode("ascii"))
        data = json.loads(raw.decode("utf-8"))
    except (binascii.Error, ValueError, UnicodeDecodeError) as exc:
        raise ValueError(f"not a valid pairing code: {exc}") from exc
    if not isinstance(data, dict):
        raise ValueError("pairing payload is not a JSON object")
    return PairingBundle(
        brain_url=_checked_url(data.get("brain_url", ""), "brain_url"),
        token=data.get("token", ""),
        glasses_id=data.get("glasses_id", ""),
        label=data.get("label", "DreamLayer"),
        relay_url=_checked_url(data.get("relay_url", ""), "relay_url"))


def connect_all(orchestrator, bundle: PairingBundle,
                http_post=None, encode_frame=None) -> dict:
    """Wire a phone (orchestrator) to a paired Brain (and note the glasses).

    Registers the Brain as the laptop tier on the orchestrator's router and
    records the glasses id. Returns a small status dict.
    """
    from .ai_brain import connect_brain
    connected_brain = bool(bundle.brain_url)
    if connected_brain:
        connect_brain(orchestrator.brain, bundle.brain_url, bundle.token,
                      http_post=http_post, encode_frame=encode_frame)
        # a paired Brain is the Mac mini tier — actually use it (leave phone-only)
        if hasattr(orchestrator, "connect_mac_mini"):
            orchestrator.connect_mac_mini(True)
        # remember where to poll for live message pop-ups
        if hasattr(orchestrator, "brain_url"):
            orchestrator.brain_url = bundle.brain_url
            orchestrator.brain_token = bundle.token
    if bundle.glasses_id:
        orchestrator.glasses_id = bundle.glasses_id
    return {"brain": connected_brain, "glasses": bool(bundle.glasses_id),
            "url": bundle.brain_url}
